### PR TITLE
util/types: StrToDate support more format control

### PR DESCRIPTION
**Pull request by tiancaiamao**
There are still many format not support yet, mostly day of year, week of year, week day, which are complex and seldom used, so I leave them for later.

### CODE REVIEW

File: util/types/time.go [hunk 1]
```diff
@@ -1855,24 +1879,48 @@ var monthAbbrev = map[string]gotime.Month{
...
+	// "%X": yearOfWeek,                 // Year for the week where Sunday is the first day of the week, numeric, four digits; used with %V
+	// "%x": yearOfWeek,                 // Year for the week, where Monday is the first day of the week, numeric, four digits; used with %v
+	// Deprecated since MySQL 5.7.5
+	// "%y": yearTwoDigits,         // Year, numeric (two digits)
+}
+
+func matchDateWithToken(t *mysqlTimeEx, date string, token string) (remain string, succ bool) {
 	if parse, ok := dateFormatParserTable[token]; ok {
-		return parse(t, date)
+		remain, succ = parse(t, date)
```
Comment (hanfei1991) [inline util/types/time.go]: Why not return directly?

File: util/types/time.go [hunk 2]
```diff
@@ -1882,18 +1930,22 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
 }
 
 func parseTwoDigits(input string) (int, bool) {
-	if len(input) < 2 {
+	return parseSomeDigits(input, 2)
```
Comment (hanfei1991) [inline util/types/time.go]: Do we need to remain the "parseTwoDigits" function?

File: util/types/time.go [hunk 3]
```diff
@@ -1759,25 +1760,47 @@ func abbrDayOfMonth(day int) string {
...
 	}
 	return true
 }
 
+// mysqlTimeEx wraps around mysqlTime and provides the context for parsing.
+// Some date format parser is not context free, for example, day of year needs
+// to know which year, it can't be interpreted alone.
+type mysqlTimeEx struct {
+	mysqlTime
+	context map[string]int
```
Comment (hanfei1991) [inline util/types/time.go]: Add comment for this map ?

File: util/types/time.go [hunk 4]
```diff
@@ -1920,7 +1972,81 @@ func minutesNumeric(t *mysqlTime, input string) (string, bool) {
 	return input[2:], true
 }
 
-func dayOfMonthNumericTwoDigits(t *mysqlTime, input string) (string, bool) {
+func time12Hour(t *mysqlTimeEx, input string) (string, bool) {
+	// hh:mm:ss AM
+	if len(input) < 10 {
```
Comment (hanfei1991) [inline util/types/time.go]: the len of "hh:mm:ss AM" is 11

File: util/types/time.go [hunk 5]
```diff
@@ -1920,7 +1972,81 @@ func minutesNumeric(t *mysqlTime, input string) (string, bool) {
...
+	return remain, true
+}
+
+func time24Hour(t *mysqlTimeEx, input string) (string, bool) {
+	// hh:mm:ss
+	if len(input) < 8 {
+		return input, false
+	}
+
+	hour, succ := parseTwoDigits(input)
```
Comment (hanfei1991) [inline util/types/time.go]: Does it allow white spaces

File: util/types/time.go (hunk 3 above)
Comment (tiancaiamao) [inline util/types/time.go]: >  mysqlTimeEx wraps around mysqlTime and provides the context for parsing

already ...

File: util/types/time.go (hunk 1 above)
Comment (tiancaiamao) [inline util/types/time.go]: oh, that's add during debugging.... I'll remove it

File: util/types/time.go (hunk 2 above)
Comment (tiancaiamao) [inline util/types/time.go]: it's used in many place, such as parsing hour, minute, minute in "00..xx" format...

File: util/types/time.go (hunk 4 above)
Comment (tiancaiamao) [inline util/types/time.go]: you don't take whitespace count into consideration before `AM`
it maybe `hh:mm:ss              AM` or `hh:mm:ssAM`

File: util/types/time.go (hunk 5 above)
Comment (tiancaiamao) [inline util/types/time.go]: No for here.

File: util/types/time.go (hunk 4 above)
Comment (hanfei1991) [inline util/types/time.go]: Then replace 10 with string("hh:mm:ssAM").Length ?

File: util/types/time.go (hunk 2 above)
Comment (hanfei1991) [inline util/types/time.go]: I think it's ok to use parseSomeDigits(input, 2) directly

File: util/types/time.go (hunk 3 above)
Comment (hanfei1991) [inline util/types/time.go]: What kind of keys will appare in this map and what are there meanings ?

File: util/types/time.go [hunk 6]
```diff
@@ -1882,18 +1930,22 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
 }
 
 func parseTwoDigits(input string) (int, bool) {
-	if len(input) < 2 {
+	return parseSomeDigits(input, 2)
+}
+
+func parseSomeDigits(input string, count int) (int, bool) {
```
Comment (hanfei1991) [inline util/types/time.go]: name `parseDigits` is more clear.

File: util/types/time.go [hunk 7]
```diff
@@ -1759,25 +1760,47 @@ func abbrDayOfMonth(day int) string {
...
+// to know which year, it can't be interpreted alone.
+type mysqlTimeEx struct {
+	mysqlTime
+	context map[string]int
+}
+
+func mysqlTimeExFix(t *mysqlTimeEx) {
+	ctx := t.context
+	if yearOfDay, ok := ctx["%j"]; ok {
+		_ = yearOfDay // TODO
```
Comment (zimulala) [inline util/types/time.go]: Complete this `TODO`.

Comment (tiancaiamao) [inline util/types/time.go]: It rely on some function do the reverse work of `Daynr`, to compute the date according to the day in a year,  but I found we haven't implement yet, and that would be another PR.

File: util/types/time.go [hunk 8]
```diff
@@ -1783,6 +1783,7 @@ func (t *Time) StrToDate(date, format string) bool {
 // to know which year, it can't be interpreted alone.
 type mysqlTimeEx struct {
 	mysqlTime
+	// key is the format char, such as `%j` `%p` and so on.
```
Comment (hanfei1991) [inline util/types/time.go]: key -> Key

File: util/types/time.go [hunk 9]
```diff
@@ -1759,25 +1760,48 @@ func abbrDayOfMonth(day int) string {
...
 	if t.check() != nil {
 		return false
 	}
 	return true
 }
 
+// mysqlTimeEx wraps around mysqlTime and provides the context for parsing.
+// Some date format parser is not context free, for example, day of year needs
+// to know which year, it can't be interpreted alone.
+type mysqlTimeEx struct {
```
Comment (coocood) [inline util/types/time.go]: Define this type doesn't make much sense.
We can use mysqlTime and context separately.

File: util/types/time.go [hunk 10]
```diff
@@ -1881,69 +1929,203 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
 	t.minute = uint8(v)
 	return input[2:], true
 }
 
-func dayOfMonthNumericTwoDigits(t *mysqlTime, input string) (string, bool) {
-	v, succ := parseTwoDigits(input)
+func time12Hour(t *mysqlTimeEx, input string) (string, bool) {
+	// hh:mm:ss AM
+	hour, succ := parseDigits(input, 2)
+	if !succ || hour > 12 || input[2] != ':' {
```
Comment (coocood) [inline util/types/time.go]: input[2] may out of range.

File: util/types/time.go [hunk 11]
```diff
@@ -1759,25 +1760,48 @@ func abbrDayOfMonth(day int) string {
...
+
+func mysqlTimeExFix(t *mysqlTimeEx) {
+	ctx := t.context
+	if yearOfDay, ok := ctx["%j"]; ok {
+		_ = yearOfDay // TODO
+	}
+	if valueAMorPm, ok := ctx["%p"]; ok {
+		if valueAMorPm == 2 {
+			t.hour += 12
+		}
```
Comment (coocood) [inline util/types/time.go]: Check if hour > 12 if has AM or PM?

File: util/types/time.go [hunk 12]
```diff
@@ -1759,25 +1760,48 @@ func abbrDayOfMonth(day int) string {
...
+// mysqlTimeEx wraps around mysqlTime and provides the context for parsing.
+// Some date format parser is not context free, for example, day of year needs
+// to know which year, it can't be interpreted alone.
+type mysqlTimeEx struct {
+	mysqlTime
+	// Key is the format char, such as `%j` `%p` and so on.
+	context map[string]int
+}
+
+func mysqlTimeExFix(t *mysqlTimeEx) {
```
Comment (coocood) [inline util/types/time.go]: May return error

File: util/types/time.go [hunk 13]
```diff
@@ -1881,69 +1929,203 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
+		return input, false
+	}
+
+	t.hour = uint8(hour)
+	t.minute = uint8(minute)
+	t.second = uint8(second)
+	return input[8:], true
+}
+
+func isAMOrPM(t *mysqlTimeEx, input string) (string, bool) {
```
Comment (zimulala) [inline util/types/time.go]: input[2:] may out of range.

File: util/types/time.go [hunk 14]
```diff
@@ -1881,69 +1929,203 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
 	return input[2:], true
 }
 
-func yearNumericFourDigits(t *mysqlTime, input string) (string, bool) {
-	if len(input) < 4 {
+var twoDigitRegex = regexp.MustCompile("^[0-9][1-9]?")
+
+// parseTwoNumeric is used for pattens 0..31 0..24 0..60 and so on.
+// It returns the parsed int, and remain data after parse.
+func parseTwoNumeric(input string) (int, string) {
```
Comment (zimulala) [inline util/types/time.go]: Ditto.

Comment (zimulala) [inline util/types/time.go]: Ditto.

File: util/types/time.go [hunk 15]
```diff
@@ -1881,69 +1929,203 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
+	}
+	v, err := strconv.ParseUint(input[:6], 10, 64)
 	if err != nil {
 		return input, false
 	}
+	t.microsecond = uint32(v)
+	return input[6:], true
+}
+
+func yearNumericFourDigits(t *mysqlTimeEx, input string) (string, bool) {
```
Comment (zimulala) [inline util/types/time.go]: Ditto.

File: util/types/time.go [hunk 16]
```diff
@@ -1881,69 +1929,203 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
+	if !succ {
+		return input, false
+	}
 	t.year = uint16(v)
 	return input[4:], true
 }
 
-func monthNumericTwoDigits(t *mysqlTime, input string) (string, bool) {
-	v, succ := parseTwoDigits(input)
+func dayOfYearThreeDigits(t *mysqlTimeEx, input string) (string, bool) {
```
Comment (zimulala) [inline util/types/time.go]: Ditto.

File: util/types/time.go [hunk 17]
```diff
@@ -1974,7 +2156,17 @@ func abbreviatedMonth(t *mysqlTime, input string) (string, bool) {
 	return input, false
 }
 
-func monthNumeric(t *mysqlTime, input string) (string, bool) {
+func fullNameMonth(t *mysqlTimeEx, input string) (string, bool) {
```
Comment (zimulala) [inline util/types/time.go]: Ditto.

File: util/types/time.go (hunk 14 above)
Comment (zimulala) [inline util/types/time.go]: I got it.

File: util/types/time.go (hunk 11 above)
Comment (tiancaiamao) [inline util/types/time.go]: actually, it will be before `StrToDate()` return...

File: util/types/time.go (hunk 14 above)
Comment (tiancaiamao) [inline util/types/time.go]: No, it's safe here...

File: util/types/time.go [hunk 18]
```diff
@@ -1775,9 +1780,27 @@ func (t *Time) StrToDate(date, format string) bool {
 	return true
 }
 
+// mysqlTimeFix fixes the mysqlTime use the values in the context.
+func mysqlTimeFix(t *mysqlTime, ctx map[string]int) error {
+	// Key of the ctx is the format char, such as `%j` `%p` and so on.
+	if yearOfDay, ok := ctx["%j"]; ok {
+		// TODO Implement the function that converts day of year to yy:mm:dd.
```
Comment (zimulala) [inline util/types/time.go]: s/TODO/TODO:

**Comment by hanfei1991** [comment]
LGTM

File: util/types/time.go [hunk 19]
```diff
@@ -1968,8 +1967,13 @@ func minutesNumeric(t *mysqlTimeEx, input string) (string, bool) {
 	return input[2:], true
 }
 
-func time12Hour(t *mysqlTimeEx, input string) (string, bool) {
+const time12HourLen = len("hh:mm:ssAM")
```
Comment (coocood) [inline util/types/time.go]: Need a space before `AM`?

File: util/types/time.go [hunk 20]
```diff
@@ -2000,7 +2004,7 @@ func time12Hour(t *mysqlTimeEx, input string) (string, bool) {
 	return remain, true
 }
 
-func time24Hour(t *mysqlTimeEx, input string) (string, bool) {
+func time24Hour(t *mysqlTime, input string, ctx map[string]int) (string, bool) {
 	// hh:mm:ss
 	if len(input) < 8 {
```
Comment (coocood) [inline util/types/time.go]: Use a constant like `time12Hour`?

File: util/types/time.go (hunk 19 above)
Comment (tiancaiamao) [inline util/types/time.go]: no, space is optional

Comment (tiancaiamao) [inline util/types/time.go]:
```
mysql> select str_to_date('10:21:25 AM', '%r');
+----------------------------------+
| str_to_date('10:21:25 AM', '%r') |
+----------------------------------+
| 10:21:25                         |
+----------------------------------+
1 row in set (0.00 sec)

mysql> select str_to_date('10:21:25AM', '%r');
+---------------------------------+
| str_to_date('10:21:25AM', '%r') |
+---------------------------------+
| 10:21:25                        |
+---------------------------------+
1 row in set (0.00 sec)

mysql> select str_to_date('10:21:25   AM', '%r');
+------------------------------------+
| str_to_date('10:21:25   AM', '%r') |
+------------------------------------+
| 10:21:25                           |
+------------------------------------+
1 row in set (0.00 sec)
```

File: util/types/time.go [hunk 21]
```diff
@@ -1775,9 +1780,27 @@ func (t *Time) StrToDate(date, format string) bool {
...
 
+// mysqlTimeFix fixes the mysqlTime use the values in the context.
+func mysqlTimeFix(t *mysqlTime, ctx map[string]int) error {
+	// Key of the ctx is the format char, such as `%j` `%p` and so on.
+	if yearOfDay, ok := ctx["%j"]; ok {
+		// TODO Implement the function that converts day of year to yy:mm:dd.
+		_ = yearOfDay
+	}
+	if valueAMorPm, ok := ctx["%p"]; ok {
+		if valueAMorPm == 2 {
```
Comment (coocood) [inline util/types/time.go]: Use a constant instead of 2

File: util/types/time.go [hunk 22]
```diff
@@ -1775,9 +1780,27 @@ func (t *Time) StrToDate(date, format string) bool {
...
+			t.hour += 12
```
Comment (coocood) [inline util/types/time.go]: 12:00 AM is 00:00 in 24 hour

File: util/types/time.go [hunk 23]
```diff
@@ -1881,69 +1928,208 @@ func matchDateWithToken(t *mysqlTime, date string, token string) (remain string,
...
 	if !succ || v >= 32 {
 		return input, false
 	}
 	t.day = uint8(v)
 	return input[2:], true
 }
 
-func yearNumericFourDigits(t *mysqlTime, input string) (string, bool) {
-	if len(input) < 4 {
+var twoDigitRegex = regexp.MustCompile("^[0-9][1-9]?")
```
Comment (coocood) [inline util/types/time.go]: The expression is incorrect.

**Comment by coocood** [comment]
LGTM
